include/onnxruntime/core/framework/sparse_tensor.h

```diff
@@ -1,72 +1,436 @@
+  /// The function allocates a single contiguous buffer and copies values and index
+  /// into it. The shape of the values is expected to be at least 3-D but may contain more
+  /// dimensions. At the very minimum it should be (num_blocks, block_size, block_size).
+  ///
+  // The shape of the index is must be at least 2-D and must contain one tuple per each of
+  // the value blocks that  were supplied to the constructor. Each index tuple is a
+  // (row, col) coordindate of the values block in a dense matrix.
+  /// </summary>
+  /// <param name="data_transfer"></param>
+  /// <param name="data_location"></param>
+  /// <param name="values_shape"></param>
+  /// <param name="values_data"></param>
+  /// <param name="indices_shape"></param>
+  /// <param name="indices_data"></param>
+  /// <returns></returns>
+  Status MakeBlockSparseData(const IDataTransfer& data_transfer,
+                             const OrtMemoryInfo& data_location,
+                             const TensorShape& values_shape, const void* values_data,
+                             const TensorShape& indices_shape, const int32_t* indices_data);
+
+  /// <summary>
+  /// Mutable data access
+  /// </summary>
+  class BlockSparseMutator {
+   public:
+    BlockSparseMutator(Tensor& values, Tensor& indices) noexcept
+        : values_(values), indices_(indices) {}
+    Tensor& Values() noexcept { return values_; }
+    Tensor& Indices() noexcept { return indices_; }
+
+   private:
+    std::reference_wrapper<Tensor> values_;
+    std::reference_wrapper<Tensor> indices_;
+  };
+
+  /// <summary>
+  /// Allocates memory for values and index and returns mutator so
+  /// data can be populated
+  /// </summary>
+  /// <param name="values_shape"></param>
+  /// <param name="indices_shape"></param>
+  /// <returns></returns>
+  BlockSparseMutator MakeBlockSparseData(const TensorShape& values_shape, const TensorShape& indices_shape);
+
+  /// <summary>
+  /// X-device copy. Destination tensor must have allocator set.
+  /// </summary>
+  /// <param name="data_transfer_manager"></param>
+  /// <param name="exec_q_id"></param>
+  /// <param name="dst_tensor"></param>
+  /// <returns></returns>
+  Status Copy(const DataTransferManager& data_transfer_manager, int exec_q_id, SparseTensor& dst_tensor) const;
+
+  /// <summary>
+  /// X-device copy. Destination tensor must have allocator set.
+  /// </summary>
+  /// <param name="dst_tensor"></param>
+  /// <returns></returns>
+  Status Copy(const IDataTransfer& data_transfer, SparseTensor& dst_tensor, int exec_q_id) const;
 
  private:
-  Tensor values_;
-  Tensor indices_;
-  TensorShape shape_;  // The shape of corresponding dense-tensor.
+  Status AllocateBuffer(int64_t buffer_size, size_t num_values);
+  void ReleaseBuffer();
+  void* IndicesStart(int64_t values_bytes);
+  const void* IndicesStart(int64_t values_bytes) const;
+  Status ValidateBlockSparseShapes(const TensorShape& values_shape, const TensorShape& index_shape) const;
```

Comment:
nit: these could call values_.SizeInBytes() internally instead of requiring the caller to pass that in.

Reply:
I feel this comment is misplaced. However, in other places, bytes are needed for 1) Tensors are not initialized yet 2) still need to verify that the user is passing the right values.